File: python/494.target_sum.py
class Solution:
    def findTargetSumWays(self, nums: List[int], S: int) -> int:
        sumAll = sum(nums)
        if S > sumAll or (S + sumAll) % 2:
            return 0
        target = (S + sumAll) // 2

        dp = [0] * (target + 1)
        dp[0] = 1

        for num in nums:
            for j in range(target, num - 1, -1):
                dp[j] = dp[j] + dp[j - num]
        return dp[-1]

作者：edelweisskoko
链接：https://leetcode-cn.com/problems/target-sum/solution/494-mu-biao-he-dong-tai-gui-hua-zhi-01be-78ll/
来源：力扣（LeetCode）
著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

File: python/494.target_sum.cpp
// solution 1 DFS 

class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int target) 
    
    {
        const int sum = std::accumulate(nums.begin(), nums.end(), 0);
        if(sum < std::abs(target)) return 0;
        int ans =0;
        dfs(nums, 0, target, ans);
        return ans;
    }
    void dfs(vector<int>& nums, int d, int target, int &ans)
    {
        if(d == nums.size())
        {
            if(target == 0)
                ans++;
            return;
        }
        dfs(nums, d+1, target - nums[d], ans);
        dfs(nums, d+1, target + nums[d], ans);
    }
};
https://www.youtube.com/watch?v=r6Wz4W1TbuI